_posts/2016-06-19-graphql-for-mobile.markdown

```diff
@@ -0,0 +1,99 @@
+
+```json
+{
+  artwork(id: "kimber-berry-as-close-to-magic-as-you-can-get") {
+    id
+    additional_information
+
+    is_price_hidden
+    is_inquireable
+  }
+}
+
+```
+
+> This will search for a [specific artwork](https://www.artsy.net/artwork/kimber-berry-as-close-to-magic-as-you-can-get), sending back the Artwork's `id`, `additional_information`, `is_price_hidden` and `is_inquireable`.
+
+It's important to note here, the data being sent _back_ is only what you ask for. This is not defined on the server as a _short_ or _embedded_ version of a model, but the specific data the client requested. When bandwidth and speed is crucial, this is the other way in which GraphQL improves the end-user experience.
+
+So, that's the two killer features:
+
+1. Coelesce Multiple Network Requests.
+2. Only Send The Data You Want.
+
+This is in stark contrast to existing API concepts, like [HAL](http://stateless.co/hal_specification.html) and [JSON-API](http://jsonapi.org) - both of which are optimised for caching, and rely on "one model, one request" types of API access. E.g. a list of Artworks would actually contain a list of hrefs instead of the model data, and you have to fetch each model as a separate request.
+
+### Using GraphQL
+
+Artsy's GraphQL server is (unsurprisingly) open-source, it's at [artsy/metaphysics](https://github.com/artsy/metaphysics). However, it's not publicly accessible, ([yet?](https://github.com/artsy/metaphysics/issues/279)). One of the coolest things about developing against a GraphQL server is that it has a built-in IDE. I can't show you ours, but I can send you to [Clay Allsop's](http://clayallsopp.com) [GraphQLHub](https://www.graphqlhub.com):
+
+[Here][hub] ( I strongly recommend pausing to open that link in a new window. Press cmd + enter to see the results )
+
+GraphQL comes with a playground for the API! It's amazing! Clay called it the ["Killer App" of GraphQL](https://medium.com/the-graphqlhub/graphiql-graphql-s-killer-app-9896242b2125#.6ht6374bq) - I'm inclined to concur. I've never had API docs this useful.
+
+{% expanded_img /images/2016-06-19-graphql-for-iOS-devs/graphiql.png Selection diagram %} 
+
+### How GraphQL Changed How We Write Native Code
+
+#### View Models
+
+Our GraphQL server is owned by the [web-practice](http://artsy.github.io/blog/2016/03/28/artsy-engineering-organization-stack/) and the mobile practice also help out occasionally. This ownership distinction is important, an API like this would normally be handled by our platform team. 
+
+Because of Metaphysics' ownership as a "front-end" product, it can contain additional information that is specific to our needs. For example, in our first example of a request to our GraphQL server we requested `id`, `additional_information`, `is_price_hidden` and `is_inquireable` - only two of these datums come from the database. The `is_` denotes a derived state in our implmentation.
```

Comment:
Also: `is_` denotes a Boolean value. There's no naming scheme or differentiation between derived fields and server returned data.